core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/container/ContainerManager.scala

```diff
@@ -360,6 +337,100 @@ object ContainerManager {
    */
   def rng(mod: Int): Int = ThreadLocalRandom.current().nextInt(mod)
 
+  // Partition messages that can use warmed containers.
+  // return: (list of messages that cannot use warmed containers, list of messages that can take advantage of warmed containers)
+  protected[container] def filterWarmedCreations(warmedContainers: Set[String],
+                                                 inProgressWarmedContainers: TrieMap[String, String],
+                                                 invokers: List[InvokerHealth],
+                                                 msgs: List[ContainerCreationMessage])(
+    implicit logging: Logging): (List[(ContainerCreationMessage, Option[Int], Option[String])],
+                                 List[(ContainerCreationMessage, Option[Int], Option[String])]) = {
+    val warmedApplied = msgs.map { msg =>
+      val warmedPrefix =
+        containerPrefix(ContainerKeys.warmedPrefix, msg.invocationNamespace, msg.action, Some(msg.revision))
+      val container = warmedContainers
+        .filter(!inProgressWarmedContainers.values.toSeq.contains(_))
+        .flatMap { container =>
+          if (container.startsWith(warmedPrefix)) {
+            logging.info(this, s"Choose a warmed container $container")
+
+            // this is required to exclude already chosen invokers
+            inProgressWarmedContainers.update(msg.creationId.asString, container)
+            Some(container)
+          } else
+            None
+        }
+
+      // chosenInvoker is supposed to have only one item
+      val chosenInvoker = container
+        .map(_.split("/").takeRight(3).apply(0))
+        // filter warmed containers in disabled invokers
+        .filter(
+          invoker =>
+            invokers
+            // filterWarmedCreations method is supposed to receive healthy invokers only but this will make sure again only healthy invokers are used.
+              .filter(invoker => invoker.status.isUsable)
+              .map(_.id.instance)
+              .contains(invoker.toInt))
+
+      if (chosenInvoker.nonEmpty) {
+        (msg, Some(chosenInvoker.head.toInt), Some(container.head))
+      } else
+        (msg, None, None)
+    }
+
+    warmedApplied.partition { item =>
+      if (item._2.nonEmpty) false
```

Comment:
If a tuple `_2.` is not empty which means there "is" a chosen invoker, this would be a warmed creation.